CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,300 @@
+{
+    badChHE_.clear();
+    badChHF_.clear();
+}
+
+void PFCandidateRecalibrator::produce(edm::Event &iEvent, const edm::EventSetup &iSetup)
+{
+    //access calogeometry
+    edm::ESHandle<CaloGeometry> calogeom;
+    iSetup.get<CaloGeometryRecord>().get(calogeom);
+    const CaloGeometry* cgeo = calogeom.product();
+    const HcalGeometry* hgeom = static_cast<const HcalGeometry*>(cgeo->getSubdetectorGeometry(DetId::Hcal,HcalForward));
+ 
+    //access PFCandidates
+    edm::Handle<std::vector<reco::PFCandidate>> pfcandidates;
+    iEvent.getByToken(pfcandidates_, pfcandidates);
+
+    int n = pfcandidates->size();
+    std::unique_ptr<std::vector<reco::PFCandidate>> copy(new std::vector<reco::PFCandidate>());
+    std::unique_ptr<std::vector<reco::PFCandidate>> discarded(new std::vector<reco::PFCandidate>());
+    copy->reserve(n);
+    std::vector<int> oldToNew(n), newToOld, badToOld; 
+    newToOld.reserve(n);
+
+    LogDebug("PFCandidateRecalibrator") << "NEW EV:" << std::endl;
+
+    //loop over PFCandidates
+    int i = -1;
+    for (const reco::PFCandidate &pf : *pfcandidates) 
+      {
+	++i;
+
+	//deal with HE
+	if( pf.particleId() == 5 && 
+	    fabs(pf.eta()) > 1.4  && fabs(pf.eta()) < 3.)
+	  {
+	    bool toKill = false;
+	    for(auto const& badIt: badChHE_)
+	      if( reco::deltaR2(pf.eta(), pf.phi(), std::get<0>(badIt), std::get<1>(badIt)) < 0.07 )
+		toKill = true;
+	    
+	    
+	    if(toKill)
+	      {
+		discarded->push_back(pf);
+		oldToNew[i] = (-discarded->size());
+		badToOld.push_back(i);
+		continue;
+	      }
+	    else
+	      {
+		copy->push_back(pf);
+		oldToNew[i] = (copy->size());
+		newToOld.push_back(i);
+	      }
+	  }
+	//deal with HF
+	else if(fabs(pf.eta()) > 3.)
+	  {
+	    const math::XYZPointF& ecalPoint = pf.positionAtECALEntrance();
+	    GlobalPoint ecalGPoint(ecalPoint.X(),ecalPoint.Y(),ecalPoint.Z());
+	    HcalDetId closestDetId(hgeom->getClosestCell(ecalGPoint));
+	    
+	    if(closestDetId.subdet() == 4)
+	      {
+		HcalDetId hDetId(closestDetId.subdet(),closestDetId.ieta(),closestDetId.iphi(),1); //depth1
+		
+		//raw*calEnergy() is the same as *calEnergy() - no corrections are done for HF
+		float longE  = pf.rawEcalEnergy() + pf.rawHcalEnergy()/2.;  //depth1
+		float shortE = pf.rawHcalEnergy()/2.;                       //depth2
+		
+		float ecalEnergy = pf.rawEcalEnergy();
+		float hcalEnergy = pf.rawHcalEnergy();
+		float totEnergy = ecalEnergy + hcalEnergy;
+		
+		bool toKill = false;
+			
+		for(auto const& badIt: badChHF_)
+		  {
+		    if ( hDetId.ieta() == std::get<0>(badIt) &&
+			 hDetId.iphi() == std::get<1>(badIt) )
+		      {
+			LogDebug("PFCandidateRecalibrator") << "==> orig en (tot,H,E): " << pf.energy() << " " << pf.rawHcalEnergy() << " " << pf.rawEcalEnergy() << std::endl;
+			if(std::get<2>(badIt) == 1) //depth1
+			  {
+			    longE *= std::get<3>(badIt);
+			    ecalEnergy = longE - shortE;
+			    totEnergy = ecalEnergy + hcalEnergy;
```

Comment:
remind me please if it's OK here to have ecalEnergy < 0.
If it has to be non-negative, some truncation logic should be added.

Reply:
good point. from looking at
```
https://github.com/cms-sw/cmssw/blob/master/RecoParticleFlow/PFProducer/src/PFAlgo.cc#L1545-L1594
```
it seems that the energy of the `PFCandidates` in HF comes straight from the `PFCluster` energy so the relevant thresholds are the ones applied at `PFRechits` and `PFCluster` level I would say.